apf/apfAggregateNumbering.cc

```diff
@@ -0,0 +1,365 @@
+#include "apfAggregateNumbering.h"
+#include "apfField.h"
+#include "apfNumberingClass.h"
+#include "apfTagData.h"
+#include "apfUserData.h"
+#include <PCU.h>
+#include <pcu_util.h>
+#include <set>
+namespace apf
+{
+  template <class T>
+  class AggregateNumberingOf : public NumberingOf<T>
+  {
+  public:
+    AggregateNumberingOf()
+      : NumberingOf<T>()
+      , nd_ids(NULL)
+      , shr(NULL)
+      , blks_per_nd(0)
+      , dofs_per_blk(0)
+      , slf(-1)
+      , strds()
+      , frsts()
+      , lcl_strd(0)
+    { }
+    void init(const char * n,
+              Mesh * m,
+              Sharing * sh,
+              FieldShape * s,
+              int blks,
+              int bs);
+    void init(Field * f, Sharing * sh, int blks, int bs);
+    void globalize();
+    virtual void getAll(MeshEntity * e, T * dat);
+    virtual T get(MeshEntity * e, int nd, int cmp);
+    virtual void set(MeshEntity*,int,int,T) {}
+    Sharing * getSharing() const { return shr; }
+    void setSharing(Sharing * s) { shr = s; }
+    int countBlocks() const { return blks_per_nd; }
+    int blockSize() const { return dofs_per_blk; }
+    int getLocalStride() const { return lcl_strd; }
+    int getScopeStride() const { return strds[slf]; }
+    int getLocalFirstDof() const { return frsts[slf]; }
+    int getStride(int peer) const { return strds[peer]; };
+    int getFirstDof(int peer) const { return frsts[peer]; }
+  private:
+    FieldDataOf<T> * nd_ids;
+    Sharing * shr;
+    int blks_per_nd;
+    int dofs_per_blk;
+    int slf;
+    DynamicArray<T> strds;
+    DynamicArray<T> frsts;
+    int lcl_strd;
+  };
+  // this assumes that the ranks returned by a sharing
+  //  do not change when the pcu comm changes
+  //  if the sharing can handle the pcu comm changing
+  //  then this isn't needed
+  class StaticSharing : public NormalSharing
+  {
+  protected:
+    Sharing * shr;
+    // original comm
+    MPI_Comm pcu_cm;
+    int pcu_rnk;
+    MPI_Group pcu_grp;
+    // new comm
+    MPI_Comm new_cm;
+    MPI_Group new_grp;
+  public:
+    StaticSharing(Mesh * m,
+                  Sharing * s,
+                  MPI_Comm ncm,
+                  MPI_Comm pcm = MPI_COMM_NULL)
+      : NormalSharing(m)
+      , shr(s)
+      , pcu_cm(pcm)
+      , pcu_rnk(-1)
+      , pcu_grp()
+      , new_cm(ncm)
+      , new_grp()
+    {
+      if(pcu_cm == MPI_COMM_NULL)
+        pcu_cm = PCU_Get_Comm();
+      new_cm = PCU_Get_Comm();
+      MPI_Comm_group(pcu_cm,&pcu_grp);
+      if(new_cm != MPI_COMM_NULL)
+        MPI_Comm_group(new_cm,&new_grp);
```

Comment:
Could we easily provide PCU wrappers for the group function and  `MPI_Group_translate_ranks`?